### PR TITLE
Clean up miscellaneous caching issues and race conditions

### PR DESCRIPTION
**Pull request by ahorner**
- Short-circuit re-association loop for newly-synchronized or internally updated records
- Cache the appropriate timestamp on record propagation
- Ensure proper scoping of cache keys for update timestamp tracking
- Add regression spec for basic race condition case

### CODE REVIEW

File: lib/restforce/db/worker.rb [hunk 1]
```diff
@@ -104,11 +113,7 @@ def track
         if tracker
           runtime = Time.now
 
-          if tracker.last_run
-            log "SYNCHRONIZING from #{tracker.last_run.iso8601}"
-          else
-            log "SYNCHRONIZING"
-          end
+          log "SYNCHRONIZING#{" from #{tracker.last_run.iso8601}" if tracker.last_run}"
```
Comment (rdunlop) [inline lib/restforce/db/worker.rb]: I don't find this clearer.

**Comment by rdunlop** [comment]
:+1:

File: lib/restforce/db/worker.rb (hunk 1 above)
Comment (ahorner) [inline lib/restforce/db/worker.rb]: Dirty secret: I did it to reduce the number of lines in the class so rubocop wouldn't complain. You're right that it doesn't read as clearly. I'll work through a few changes to reduce complexity where it actually matters.

Comment (ahorner) [inline lib/restforce/db/worker.rb]: I found some meaningful places to reduce the amount of code in the class; should be good for another look. Thanks for keeping me honest.

**Comment by rdunlop** [comment]
:+1:

**Pull request merged**